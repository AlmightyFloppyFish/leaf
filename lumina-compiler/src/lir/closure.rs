//! Lowering for Closures and partial application

use super::{ssa, FuncLower, Function, MonoFunc, MonoType, MonoTypeKey, Value};
use crate::prelude::*;

impl<'a> FuncLower<'a> {
    pub fn partially_applicate_func(&mut self, target: MonoFunc, given: Vec<lir::Value>) -> Value {
        info!("partially applicating {target} with {given:?}");
        let given_types = given.iter().map(|v| self.type_of_value(*v)).collect();
        let vtable = self.partial_application_vtable(target, given_types);
        let data = self.ssa().construct(given, vtable.object.into());
        self.construct_dyn_object(&vtable, data.value())
    }

    // Allocates and pairs opaque pointer for data to a vtable
    pub fn construct_dyn_object(&mut self, vtable: &VTable, data: Value) -> Value {
        let data_ptr = self.heap_alloc(data, vtable.data_type.clone());

        let vtable_ptr = self.ssa().val_to_ref(vtable.val, vtable.vtable_type.into());

        self.ssa()
            .construct(vec![data_ptr, vtable_ptr.into()], vtable.object.into())
            .value()
    }

    // MyTrait:target (impltor as MyTrait) 40
    //
    // lowers to
    //
    // { data = alloc impltor, vtable }
    // where
    //   val vtable = [
    //     fn _ ptr a =
    //       let object = deref ptr in MyTrait:target object a
    //   ]
    pub fn trait_impl_vtable(
        &mut self,
        trait_: M<key::Trait>,
        impltor: MonoType,
        methods: Map<key::Method, MonoFunc>,
    ) -> VTable {
        let module = self.current.origin.module();

        let data_type = impltor;

        let Some(param_pos) = self.mir.trait_objects[trait_].as_ref() else {
            panic!("{trait_} is not object safe");
        };
        let vtable_type =
            self.new_vtable_type(VTableSelf::Method(param_pos), methods.values().copied());
        let object_type = self.new_object_type(trait_, vtable_type);

        if let Some(val) = self.lir.vtables.get(&object_type).copied() {
            return VTable { val, object: object_type, data_type, vtable_type };
        }

        let forwarding_funcs = methods
            .into_iter()
            .map(|(method, target)| {
                // Get the function typing by accessing the function pointer in the vtable
                let vtable_call_field = self
                    .lir
                    .types
                    .types
                    .type_of_field(vtable_type, key::RecordField(method.0));
                let (vtable_func_ptypes, ret) = vtable_call_field.as_fnptr();

                let mut blocks = ssa::Blocks::new(vtable_func_ptypes.iter().cloned().collect());

                let self_bparam = lir::BlockParam(param_pos[method].0);

                // Dereference the opaque data
                let data = blocks.deref(Value::BlockParam(self_bparam), data_type.clone());

                // Forward the parameters
                let params = (0..vtable_func_ptypes.len())
                    .map(|i| {
                        if self_bparam.0 == i as u32 {
                            data.value()
                        } else {
                            Value::BlockParam(lir::BlockParam(i as u32))
                        }
                    })
                    .collect();

                // Call the target
                blocks.jump(target, params);

                // Define the forwarding function
                self.lir.push_function(
                    format!("VTable_{target}_{object_type}"),
                    blocks,
                    ret.clone(),
                )
            })
            .collect::<Vec<_>>();

        let val = self.create_vtable_global(
            module,
            data_type.clone(),
            vtable_type,
            forwarding_funcs.into_iter(),
        );
        self.lir.vtables.insert(vtable_type, val);

        VTable { val, object: object_type, data_type, vtable_type }
    }

    // f #(target a b) // `target` takes 3 parameters
    //
    // lowers to
    //
    // { data = alloc (a, b), vtable } : fn(c -> R)
    // where           ^^^^ 'given: Vec<MonoType>'
    //   val vtable = [
    //     fn _ ptr c as *u8, c -> R =
    //       let (a, b) = deref ptr in target a b c
    //   ]
    pub fn partial_application_vtable(&mut self, target: MonoFunc, given: Vec<MonoType>) -> VTable {
        // TODO: currently these vals are declared in the module where they're first used.
        // That's a bit odd. We should probably make it so that it uses `target.module()`
        let module = self.current.origin.module();
        let given_count = given.len();

        let vtable_type = self.new_vtable_type(
            VTableSelf::PartialApplication(given_count),
            std::iter::once(target),
        );
        let data_type = self.lir.types.get_or_make_tuple(given);
        let object_type = self.new_object_type(self.info.closure, vtable_type);

        if let Some(val) = self.lir.vtables.get(&vtable_type).copied() {
            return VTable {
                val,
                object: object_type.into(),
                data_type: data_type.into(),
                vtable_type,
            };
        }

        let vtable_call_field = self
            .lir
            .types
            .types
            .type_of_field(vtable_type, key::RecordField(0));
        let (vtable_func_ptypes, ret) = vtable_call_field.as_fnptr();

        let target_func = &self.lir.functions[target];

        // Create the function that's put in the vtable
        let forwarding_func = {
            let mut blocks = ssa::Blocks::new(vtable_func_ptypes.iter().cloned().collect());

            let mut params = if given_count == 0 {
                vec![]
            } else {
                // Dereference the opaque data
                let data = blocks.deref(Value::BlockParam(lir::BlockParam(0)), data_type.into());

                let rdata = &self.lir.types.types[data_type];
                assert!(rdata.autoboxed.is_empty());

                // Flatten the values by dereferencing the opaque data into parameters for target
                self.lir.types.types[data_type]
                    .fields
                    .iter()
                    .map(|(field, ty)| {
                        blocks
                            .field(data.value(), data_type, field, ty.clone())
                            .value()
                    })
                    .collect::<Vec<_>>()
            };

            // Append all parameters after the flattened values as additional params for target
            for (i, _) in target_func
                .blocks
                .params(lir::Block::entry())
                .iter()
                .skip(given_count)
                .enumerate()
            {
                let pid = lir::BlockParam(i as u32 + 1); // offset to skip the capture data
                let additional = Value::BlockParam(pid);
                params.push(additional);
            }

            // Call the target
            blocks.jump(target, params);

            // Define the forwarding func
            self.lir.push_function(
                format!("VTable_({})_Closure", self.lir.functions[target].symbol),
                blocks,
                ret.clone(),
            )
        };

        let fns = std::iter::once(forwarding_func);
        let val = self.create_vtable_global(module, data_type.into(), vtable_type, fns);
        self.lir.vtables.insert(vtable_type, val);

        VTable {
            val,
            object: object_type.into(),
            data_type: data_type.into(),
            vtable_type,
        }
    }

    // Generate the initialiser and global for the vtable
    fn create_vtable_global(
        &mut self,
        module: key::Module,
        data_type: MonoType,
        vtable: MonoTypeKey,
        fns: impl Iterator<Item = MonoFunc> + Clone,
    ) -> M<key::Val> {
        let vtable_val_initialiser = {
            let mut blocks = ssa::Blocks::new(Map::new());

            let fn_pointers = fns.clone().map(Value::FuncPtr).collect();
            let v = blocks.construct(fn_pointers, vtable.into());
            blocks.return_(v.value());

            let init_name = format!(
                "VTable_{}_{}^{}_initialiser",
                self.ty_symbol(&data_type),
                fns.format(","),
                self.ty_symbol(&MonoType::Monomorphised(vtable)),
            );
            self.lir.push_function(init_name, blocks, vtable.into())
        };

        let val = self.lir.vals.push(module, vtable.into());
        self.lir
            .val_initialisers
            .insert(val, vtable_val_initialiser);

        val
    }

    // Construct the vtable type by iterating the functions to be included while substituting
    // `self` for `*u8`. If the position of `self` isn't specified then assume partial application
    // and substitute `N` parameters with a single `*u8`.
    fn new_vtable_type<'s>(
        &mut self,
        param: VTableSelf<'s>,
        methods: impl IntoIterator<Item = MonoFunc>,
    ) -> MonoTypeKey {
        // Using tuples here could be considered dangerous as it relies on the `self.lir.vtables`
        // being used.
        let fields = methods
            .into_iter()
            .enumerate()
            .map(|(i, mfunc)| {
                let func = &self.lir.functions[mfunc];

                let target_params = func.blocks.params(lir::Block::entry());

                let params = match param {
                    VTableSelf::PartialApplication(count) => {
                        let mut params = vec![MonoType::u8_pointer()];
                        params.extend(target_params.values().skip(count).cloned());
                        params
                    }
                    VTableSelf::Method(methods) => {
                        let mut params = target_params.values().cloned().collect::<Vec<_>>();
                        params[methods[key::Method(i as u32)].0 as usize] = MonoType::u8_pointer();
                        params
                    }
                };

                MonoType::fn_pointer(params, func.returns.clone())
            })
            .collect();

        self.lir.types.get_or_make_tuple(fields)
    }

    // Create the object struct which is an opaque pointer attached to the vtable pointer
    fn new_object_type(&mut self, trait_: M<key::Trait>, vtable: MonoTypeKey) -> MonoTypeKey {
        self.lir.types.get_or_make_record(
            trait_,
            vec![MonoType::u8_pointer(), MonoType::pointer(vtable.into())],
        )
    }
}

#[derive(Clone)]
pub struct VTable {
    val: M<key::Val>,    // global for vtable
    data_type: MonoType, // the real type behind `*u8`
    vtable_type: MonoTypeKey,
    object: MonoTypeKey, // *u8+vtable
}

enum VTableSelf<'a> {
    PartialApplication(usize),
    Method(&'a Map<key::Method, key::Param>),
}
