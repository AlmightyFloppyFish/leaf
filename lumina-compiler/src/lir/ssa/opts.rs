//! Optimizations that are more specific to Lumina and more appropriate for a higher-level IR than
//! our backends.

use crate::lir::{
    ssa::rewrite::for_entry_mut, ty_fmt, Block, BlockJump, Entry, Function, MonoFunc, Value, LIR,
    SSA, V,
};
use lumina_collections::KeysIter;
use smallvec::SmallVec;
use std::mem::take;
use tracing::{info, info_span, trace};

pub const ENABLE_OPTS: bool = true;

impl LIR {
    pub fn perform_optimizations(&mut self) {
        let mut opt = Optimizer { lir: self };
        opt.optimizations();
    }
}

struct Optimizer<'a> {
    lir: &'a mut LIR,
}

type Changed = bool;

impl<'a> Optimizer<'a> {
    fn optimizations(&mut self) {
        if !ENABLE_OPTS {
            return;
        }

        self.block_opts();
        self.func_opts();
    }

    fn block_opts(&mut self) {
        const FUEL: usize = 1000;

        for fkey in self.lir.functions.keys() {
            let _span = info_span!(
                "running block optimizations",
                entity = self.lir.functions[fkey].symbol.clone(),
            );
            let _handle = _span.enter();
            info!(
                "BEFORE OPTIMIZATIONS:\n{}",
                self.lir
                    .mono
                    .fmt(&self.lir.functions[fkey])
                    .fns(&self.lir.functions)
            );

            let any_change = self.with_fuel(FUEL, |lir| {
                let func = &mut lir.functions[fkey];
                let changed = func
                    .ssa
                    .blocks
                    .keys()
                    .any(|block| block_opt_iter(func, fkey, block));

                if func.symbol.contains("tuple::show::mfunc96") {
                    if changed {
                        info!(
                            "MIDDLE OPTIMIZATION:\n{}",
                            ty_fmt(&lir.mono.types, &lir.functions[fkey]).fns(&lir.functions)
                        );
                    }
                }
                changed
            });

            if any_change {
                trace!(
                    "post-optimizations:\n{}",
                    self.lir.mono.fmt(&self.lir.functions[fkey])
                );
            }
        }
    }

    fn func_opts(&mut self) {
        const FUEL: usize = 1000;

        for fkey in self.lir.functions.keys() {
            self.with_fuel(FUEL, |lir| {
                let ssa = &lir.functions[fkey].ssa;
                ssa.blocks
                    .keys()
                    .any(|block| func_opt_iter(lir, fkey, block))
            });
        }
    }

    fn with_fuel(&mut self, mut fuel: usize, mut f: impl FnMut(&'_ mut LIR) -> Changed) -> Changed {
        let mut changed = true;
        let mut any_change = false;

        while fuel != 0 && changed {
            changed = f(self.lir);
            any_change |= changed;

            fuel -= 1;
        }

        any_change
    }
}

fn block_opt_iter(func: &mut Function, fkey: MonoFunc, block: Block) -> Changed {
    if func.ssa.blocks[block].predecessors == 0 {
        info!("purging {block}");
        func.ssa.purge_block(block);
        return true;
    }

    let start = func.ssa.blocks[block].start;

    for v in KeysIter::range(start, usize::MAX) {
        if !func.ssa.ventries.has(v) {
            break;
        }

        match &func.ssa.ventries[v] {
            // block0():
            //  v1 = 0
            //  jump block1(v1)
            //
            // block1(v2): // predecessors = 1
            //  return v2
            // -----
            // block0():
            //  v1 = 0
            //  return v1
            //  ...
            Entry::JmpBlock(jump)
                if func.ssa.blocks[jump.id].predecessors == 1 && jump.id == Block(block.0 + 1) =>
            {
                info!("inlining {} in {}", jump.id, block);
                let jump = jump.clone();

                for vv in KeysIter::range(v, usize::MAX) {
                    if !func.ssa.ventries.has(vv) {
                        break;
                    }

                    for_entry_mut(&mut func.ssa.ventries[vv], |vv| {
                        if vv.0 > v.0 && vv.0 <= v.0 + jump.params.len() as u32 {
                            let start = v.0 + 1;
                            let i = vv.0 - start;
                            let new = jump.params[i as usize];
                            info!("at {v}, {vv} detected as parameter {i}, rewriting to {new}");
                            new
                        } else {
                            vv.value()
                        }
                    });
                }

                // Remove the jump instruction, and parameter declarations of the block we'll remove
                func.ssa.blocks.as_mut_vec().remove(jump.id.0 as usize);
                func.ssa.delete_range(v, 1 + jump.params.len());
                func.ssa.for_each_block(V(0), |b| {
                    if b.0 > block.0 {
                        b.0 -= 1;
                    }
                });

                return true;
            }
            // Substitute tail call to self with a jump to the entry block instead
            Entry::JmpFunc(mfunc, _) if *mfunc == fkey => {
                info!("substiuting call to self with jump to entry block");
                let Entry::JmpFunc(_, params) = &mut func.ssa.ventries[v] else {
                    unreachable!()
                };
                let jump = BlockJump::new(Block::entry(), take(params));
                func.ssa.ventries[v] = Entry::JmpBlock(jump);
            }
            Entry::Select { on_true, on_false, .. } => {
                let new_on_true = try_inline_block_rejump(&func.ssa, on_true);
                let new_on_false = try_inline_block_rejump(&func.ssa, on_false);

                let Entry::Select { on_true, on_false, .. } = &mut func.ssa.ventries[v] else {
                    unreachable!()
                };

                let mut changed = false;
                if let Some(jump) = new_on_true {
                    changed = true;
                    info!("{} => {}", on_true.id, jump.id);
                    func.ssa.blocks[on_true.id].predecessors -= 1;
                    *on_true = jump;
                }
                if let Some(jump) = new_on_false {
                    changed = true;
                    info!("{} => {}", on_false.id, jump.id);
                    func.ssa.blocks[on_false.id].predecessors -= 1;
                    *on_false = jump;
                }

                return changed;
            }

            // v0 = call mfunc0()
            // return v0
            //   ----
            // jump mfunc0()
            Entry::CallStatic(..)
                if func.ssa.ventries[V(v.0 + 1)] == Entry::Return(Value::V(v)) =>
            {
                info!("substituting call+ret into tail call");

                let Entry::CallStatic(mfunc, params) = &mut func.ssa.ventries[v] else {
                    unreachable!();
                };

                let params = take(params);
                func.ssa.ventries[v] = Entry::JmpFunc(*mfunc, params);
                func.ssa.delete_range(V(v.0 + 1), 1);

                return true;
            }
            Entry::JmpTable(_, blocks) => {
                let rejumps = blocks
                    .iter()
                    .map(|block| {
                        let jmp = BlockJump::new(*block, vec![]);
                        try_inline_block_rejump(&func.ssa, &jmp)
                    })
                    .collect::<SmallVec<[_; 5]>>();

                let Entry::JmpTable(_, blocks) = &mut func.ssa.ventries[v] else {
                    unreachable!();
                };

                let changed = !rejumps.is_empty();

                for (rejump, jump) in rejumps.into_iter().zip(blocks) {
                    if let Some(new) = rejump {
                        *jump = new.id;
                    }
                }

                return changed;
            }
            _ => {}
        }
    }

    false
}

fn try_inline_block_rejump(ssa: &SSA, ijump: &BlockJump) -> Option<BlockJump> {
    let binfo = ssa.block_info(ijump.id);
    if binfo.binds == 0 {
        if let Entry::JmpBlock(jump) = binfo.tail {
            if ssa.parameters_are_local(ijump.id) {
                // Substitute the parameters in iblock with the parameters given to the
                // jump to iblock, to be used for the jump to the block jumped to by iblock.
                let params = jump
                    .params
                    .iter()
                    .map(|p| match p {
                        Value::V(v) => {
                            if let Some(i) = binfo.is_param(*v) {
                                ijump.params[i]
                            } else {
                                *p
                            }
                        }
                        _ => *p,
                    })
                    .collect();

                return Some(BlockJump::new(jump.id, params));
            }
        }
    }

    None
}

fn func_opt_iter(lir: &mut LIR, fkey: MonoFunc, block: Block) -> Changed {
    false
}

impl Entry {
    pub fn is_terminator(&self) -> bool {
        match self {
            Entry::JmpFunc(..)
            | Entry::JmpBlock(..)
            | Entry::Return(..)
            | Entry::Select { .. }
            | Entry::JmpTable(..)
            | Entry::Trap(..) => true,
            _ => false,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::debuginfo::Item;
    use crate::lir::{ty_fmt, MonoType};
    use lumina_key as key;
    use lumina_key::{Map, M};
    use tracing::info;

    fn test_block_opts(name: &str, ssa: SSA) -> (String, String) {
        let mut fuel = 5;

        let item = Item::Defined(M(key::Module::from(0), key::Func::from(0)));
        let mut func = Function::new(name.into(), item, ssa, MonoType::u(8), 1);

        let mut changed = true;

        let types = Map::new();
        lumina_util::enable_highlighting(false);
        let before = format!("BEFORE:\n{}", ty_fmt(&types, &func.ssa));

        while fuel != 0 && changed {
            changed = func
                .ssa
                .blocks
                .keys()
                .any(|block| block_opt_iter(&mut func, MonoFunc(0), block));

            if changed {
                info!("MIDDLE OPTIMIZATION:\n{}", ty_fmt(&types, &func));
            }

            fuel -= 1;
        }

        let after = format!("AFTER:\n{}", ty_fmt(&types, &func.ssa));

        (before, after)
    }

    #[test]
    fn inline_block() {
        lumina_util::test_logger();

        let mut ssa = SSA::new();

        let block = [Block::entry(), ssa.new_block()];

        let v0 = ssa.add_block_param(block[0], MonoType::u(0));
        let v1 = ssa.add(v0.value(), Value::u(1, 1), MonoType::u(1));
        let _v2 = ssa.jump(block[1], vec![v1]);

        ssa.switch_to_block(block[1]);
        let v3 = ssa.add_block_param(block[1], MonoType::u(3));
        let v4 = ssa.construct(vec![v0.value(), v1, v1, v3.value()], MonoType::u(4));
        let _v5 = ssa.return_(v4);

        let (before, after) = test_block_opts("inline_block", ssa);
        insta::assert_snapshot!(format!("{before}\n{after}"));
    }

    #[test]
    fn tricky() {
        lumina_util::test_logger();

        let mut ssa = SSA::new();

        let block = [Block::entry(), ssa.new_block(), ssa.new_block()];
        ssa.select(
            Value::bool(true),
            [
                (block[1], vec![Value::u(1, 8)]),
                (block[1], vec![Value::u(2, 8)]),
            ],
        );

        ssa.switch_to_block(block[1]);
        let p = ssa.add_block_param(block[1], MonoType::u(8));
        ssa.jump(block[2], vec![]);

        ssa.switch_to_block(block[2]);
        ssa.return_(p.value());

        let (before, after) = test_block_opts("inline_block", ssa);
        insta::assert_snapshot!(format!("{before}\n{after}"));
    }
}
