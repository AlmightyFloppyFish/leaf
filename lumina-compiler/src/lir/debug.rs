use super::*;
use derive_new::new;
use ssa::Entry;

// Scans the LIR or obvious errors
#[derive(new)]
pub struct Debugger<'a> {
    lir: &'a LIR,
    mir: &'a mir::MIR,
    #[new(value = "MonoFunc::from_u32(0)")]
    mfunc: MonoFunc,
}

impl<'a> Debugger<'a> {
    pub fn run(mut self) {
        for (mfunc, func) in self.lir.functions.iter() {
            self.mfunc = mfunc;

            let _span = info_span!("running LIR debugger", entity = func.symbol);
            let _handle = _span.enter();

            for block in func.blocks.blocks() {
                for v in func.blocks.entries(block) {
                    let entry = func.blocks.entry_of(v);
                    info!("{v} = {entry}");
                    let exp = func.blocks.type_of(v);
                    self.entry(exp, entry)
                }

                let flow = func.blocks.flow_of(block);
                self.flow(flow);
            }
        }
    }

    fn tfmt<T>(&self, v: T) -> MonoFormatter<'_, T> {
        MonoFormatter { types: &self.lir.mono.types, v }
    }

    fn tcheck(
        &self,
        mut given: impl Iterator<Item = MonoType>,
        mut expected: impl Iterator<Item = &'a MonoType>,
    ) {
        for (g, e) in (&mut given).zip(&mut expected) {
            info!(" Â· {} {} {}", self.tfmt(&g), "=?".symbol(), self.tfmt(e));
            self.check(&g, e);
        }

        assert_eq!(given.next(), None);
        assert_eq!(expected.next(), None);
    }

    #[track_caller]
    fn check(&self, got: &MonoType, exp: &MonoType) {
        if got != exp {
            dbg!(&got, &exp);
            panic!(
                "LIR type mismatch post-monomorphisation: {} != {}",
                self.tfmt(got),
                self.tfmt(exp)
            );
        }
    }

    fn flow(&mut self, flow: &ControlFlow) {
        match flow {
            ControlFlow::JmpFunc(mfunc, params) => {
                let expected = self.lir.functions[*mfunc]
                    .blocks
                    .param_types(ssa::Block::entry());

                self.params(params, expected);
            }
            ControlFlow::JmpBlock(block, params) => {
                let expected = self.lir.functions[self.mfunc].blocks.param_types(*block);
                self.params(params, expected)
            }
            ControlFlow::Unreachable => {}
            ControlFlow::Empty => {}
            ControlFlow::Return(v) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                let exp = &self.lir.functions[self.mfunc].returns;
                self.check(&ty, exp);
            }
            ControlFlow::Select { value, .. } => {
                let ty = self.lir.type_of_value(self.mfunc, *value);
                self.check(&ty, &MonoType::bool());
            }
            ControlFlow::JmpTable(v, _, _) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                self.as_int(&ty, "jump table");
            }
        }
    }

    fn params(&self, given: &[Value], expected: impl Iterator<Item = &'a MonoType>) {
        let given = given.iter().map(|v| self.lir.type_of_value(self.mfunc, *v));
        self.tcheck(given, expected);
    }

    fn entry(&mut self, exp: &MonoType, entry: &Entry) {
        match entry {
            Entry::CallStatic(mfunc, params) => {
                let expected = self.lir.functions[*mfunc]
                    .blocks
                    .param_types(ssa::Block::entry());

                self.params(params, expected);

                assert_eq!(exp, &self.lir.functions[*mfunc].returns)
            }
            Entry::Transmute(v) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                let size = self.lir.mono.types.size_of(&ty);
                let esize = self.lir.mono.types.size_of(exp);
                assert_eq!(
                    size, esize,
                    "both sides of transmute need to be of equal size: {size} != {esize}"
                );
            }
            Entry::CallExtern(fkey, params) => {
                let extern_ = &self.lir.extern_funcs[fkey];
                self.params(params, extern_.params.iter());
                assert_eq!(exp, &extern_.returns);
            }
            Entry::CallValue(to_call, params) => {
                let called = self.lir.type_of_value(self.mfunc, *to_call);
                match called {
                    MonoType::FnPointer(ptypes, ret) => {
                        self.params(params, ptypes.iter());
                        assert_eq!(exp, &*ret);
                    }
                    MonoType::Monomorphised(mkey)
                        if self
                            .lir
                            .mono
                            .types
                            .as_trait_object(mkey)
                            .map(|key| self.mir.name_of_type(key))
                            == Some("Closure") =>
                    {
                        let ty = self.lir.mono.types.as_closure_get_fnptr(mkey);
                        let (ptypes, ret) = ty.as_fnptr();
                        self.params(params, ptypes.iter());
                        assert_eq!(exp, ret);
                    }
                    ty => panic!(
                        "CallValue for non-closure or non-fnpointer: {}",
                        self.tfmt(&ty)
                    ),
                }
            }
            Entry::Copy(v) => assert_eq!(exp, &self.lir.type_of_value(self.mfunc, *v)),
            Entry::Construct(values) => match exp {
                MonoType::SumDataCast { largest } => {
                    let size = values
                        .iter()
                        .map(|v| {
                            self.lir
                                .mono
                                .types
                                .size_of(&self.lir.type_of_value(self.mfunc, *v))
                        })
                        .sum::<u32>();
                    assert!(size <= *largest);
                    todo!();
                }
                MonoType::Monomorphised(mk) => {
                    let fields = self.lir.mono.fields(*mk);
                    self.params(
                        values,
                        fields.map(|field| &self.lir.mono.types[*mk].fields[field]),
                    );
                }
                ty => panic!("cannot construct non-record: {}", self.tfmt(ty)),
            },
            Entry::RefStaticVal(val) => {
                let ty = &self.lir.vals[*val];
                assert_eq!(exp, ty);
            }
            Entry::Field { of, key, field } => {
                assert_eq!(self.lir.type_of_value(self.mfunc, *of).as_key(), *key);
                let ty = &self.lir.mono.types[*key].fields[*field];
                assert_eq!(exp, ty);
            }
            Entry::SumField { of, offset } => {
                let of = self.lir.type_of_value(self.mfunc, *of);
                match of {
                    MonoType::SumDataCast { largest } => assert!(offset.0 < largest),
                    _ => panic!("SumField of non-opaque sum data: {}", self.tfmt(&of)),
                }
            }
            Entry::IntCmpInclusive(lhs, _, rhs, _)
            | Entry::IntAdd(lhs, rhs)
            | Entry::IntSub(lhs, rhs)
            | Entry::IntMul(lhs, rhs)
            | Entry::IntDiv(lhs, rhs) => {
                let [lhs, rhs] = [lhs, rhs].map(|v| self.lir.type_of_value(self.mfunc, *v));
                assert_eq!(lhs, rhs, "{} != {}", self.tfmt(&lhs), self.tfmt(&rhs));
                self.as_int(&lhs, "div");
            }
            Entry::Reduce(v) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                self.as_int(&ty, "reduce");
            }
            Entry::ExtendSigned(v) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                assert!(self.as_int(&ty, "extend").signed);
            }
            Entry::ExtendUnsigned(v) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                assert!(!self.as_int(&ty, "extend").signed);
            }
            Entry::IntToFloat(v, size) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                assert_eq!(*size, self.as_int(&ty, "cast"));
                self.as_float(exp, "cast");
            }
            Entry::FloatToInt(v, size) => {
                let ty = self.lir.type_of_value(self.mfunc, *v);
                self.as_float(&ty, "cast");
                assert_eq!(*size, self.as_int(exp, "cast"));
            }
            Entry::BitAnd(values) => values
                .map(|v| self.lir.type_of_value(self.mfunc, v))
                .iter()
                .chain(std::iter::once(exp))
                .for_each(|v| assert_eq!(self.as_int(v, "bitand").bits(), 8)),
            Entry::BlockParam(v) => {
                let ty = self.lir.functions[self.mfunc].blocks.type_of(*v);
                assert_eq!(exp, ty);
            }
            Entry::Alloc { .. } => {}
            Entry::Dealloc { ptr } => {
                let ty = self.lir.type_of_value(self.mfunc, *ptr);
                self.as_ptr(&ty);
                self.as_unit(exp);
            }
            Entry::WritePtr { ptr, value } => {
                let ty = self.lir.type_of_value(self.mfunc, *ptr);
                let inner = self.as_ptr(&ty);
                let ty = self.lir.type_of_value(self.mfunc, *value);
                assert_eq!(ty, *inner);
                self.as_unit(exp);
            }
            Entry::Deref(ptr) => {
                // TODO: I think our casts are currently implicit for pointers. Wwe should probably change that?
                //
                // or no, they probably occur in the intcast stuff?
                let ty = self.lir.type_of_value(self.mfunc, *ptr);
                let inner = self.as_ptr(&ty);
                assert_eq!(exp, inner);
            }
        }
    }

    #[track_caller]
    fn as_unit(&self, ty: &MonoType) {
        assert_eq!(self.lir.mono.types.size_of(ty), 0, "non-unit");
    }

    #[track_caller]
    fn as_int(&self, ty: &MonoType, ctx: &str) -> IntSize {
        match ty {
            MonoType::Int(size) => *size,
            _ => panic!("non-int {ctx}: {}", self.tfmt(ty)),
        }
    }

    #[track_caller]
    fn as_float(&self, ty: &MonoType, ctx: &str) {
        match ty {
            MonoType::Float => {}
            _ => panic!("non-float {ctx}: {}", self.tfmt(ty)),
        }
    }

    #[track_caller]
    fn as_ptr<'t>(&self, ty: &'t MonoType) -> &'t MonoType {
        match ty {
            MonoType::Pointer(inner) => inner,
            _ => panic!("dealloc of non-pointer: {}", self.tfmt(ty)),
        }
    }
}
