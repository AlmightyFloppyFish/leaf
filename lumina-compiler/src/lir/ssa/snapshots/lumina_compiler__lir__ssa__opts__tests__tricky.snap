---
source: lumina-compiler/src/lir/ssa/opts.rs
expression: "format!(\"{before}\\n{after}\")"
---
BEFORE:

block0(): predecessors=1
  select 1
  | true  -> block1(1)
  | false -> block1(2) // v0

block1(v1: u8): predecessors=2
  jump block2() // v2

block2(): predecessors=1
  return v1 // v3

AFTER:

block0(): predecessors=1
  select 1
  | true  -> block1(1)
  | false -> block1(2) // v0

block1(v1: u8): predecessors=2
  return v1 // v2
